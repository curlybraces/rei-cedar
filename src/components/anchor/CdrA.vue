<template>
  <a
    :class="modifierClass"
    :target="target"
    :rel="newRel"
    :href="href"
  >
    <!-- @slot innerHTML on the inside of the anchor component -->
    <slot>Link Text</slot>
  </a>
</template>

<script>
import modifier from 'mixinsdir/modifier';

/**
 * Cedar 2 component for anchor.
 *
 * <span class="modifiers">Modifiers</span>
 *
 * Modifiers can be combined 1 from each grouping.
 * {contrast, overlay, standalone} **OR** button
 * Use of the 'button' modifier depends on including the css for CdrButton.
 * A modifier list that contains 'button' exposes the same modifiers as CdrButton and the other modifiers shouldn't be combined with it.
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrA',
  mixins: [modifier],
  props: {
    /** @ignore */
    target: String,
    /** @ignore */
    rel: String,
    href: {
      type: String,
      default: '#',
    },
  },
  computed: {
    baseClass() {
      const modifiers = this.modifier ? this.modifier.split(' ') : [];
      return modifiers.indexOf('button') >= 0 ? 'cdr-button' : 'cdr-link';
    },
    newRel() {
      if (this.target === '_blank') {
        return this.rel || 'noopener noreferrer';
      }
      return this.rel;
    },
  },
};
</script>

<style theme="default">
  @import 'cssdir/settings/_index.pcss';
  @import './styles/vars/CdrA.vars.pcss';
  @import './styles/CdrA.pcss';
</style>

