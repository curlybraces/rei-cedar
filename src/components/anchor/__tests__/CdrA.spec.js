import { shallowMount } from '@vue/test-utils';
import { createRenderer } from 'vue-server-renderer'
import CdrA from 'componentsdir/anchor/CdrA';

describe('CdrA.vue', () => {
  it('outputs the same HTML', () => {
    const wrapper = shallowMount(CdrA);
    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    });
  });

  it('renders an anchor', () => {
    const wrapper = shallowMount(CdrA);
    expect(wrapper.is('a')).toBe(true);
  });

  it('sets target attr correctly', () => {
    const wrapper = shallowMount(CdrA, {
      propsData: {
        target: '_self',
      },
    });
    expect(wrapper.attributes().target).toBe('_self');
  });

  it('sets rel attr correctly', () => {
    const wrapper = shallowMount(CdrA, {
      propsData: {
        rel: 'nofollow',
      },
    });
    expect(wrapper.attributes().rel).toBe('nofollow');
  });

  it('computes target="_blank" rel attr correctly', () => {
    const wrapper = shallowMount(CdrA, {
      propsData: {
        target: '_blank',
      },
    });
    expect(wrapper.attributes().rel).toBe('noopener noreferrer');
  });

  it('computes base class correctly', () => {
    const wrapper = shallowMount(CdrA, {
      propsData: {
        modifier: 'primary',
      },
    });
    expect(wrapper.classes()).toContain('cdr-link');
  });

  it('computes link as button base class correctly', () => {
    const wrapper = shallowMount(CdrA, {
      propsData: {
        modifier: 'primary, button',
      },
    });
    expect(wrapper.classes()).toContain('cdr-button');
  });
});