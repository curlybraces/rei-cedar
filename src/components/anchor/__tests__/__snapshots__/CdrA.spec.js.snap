// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrA.vue outputs the same HTML 1`] = `<a class="cdr-link"></a>`;
